jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -5,80 +5,304 @@
+          .createConsumer(settings.selector)
+          .onComplete {
+            case Success(consumer) =>
+              consumer.setMessageListener(new MessageListener {
+                def onMessage(message: Message): Unit = {
+                  backpressure.acquire()
+                  handleMessage.invoke(message)
+                }
+              })
+            case Failure(e) =>
+              fail.invoke(e)
+          }
+    }
+
+    (logic, logic.killSwitch)
+  }
+}
+
+final class JmsAckSourceStage(settings: JmsSourceSettings)
+    extends GraphStageWithMaterializedValue[SourceShape[AckEnvelope], KillSwitch] {
+
+  private val out = Outlet[AckEnvelope]("JmsSource.out")
+
+  override def shape: SourceShape[AckEnvelope] = SourceShape[AckEnvelope](out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, KillSwitch) = {
+
+    val logic = new SourceStageLogic[AckEnvelope](shape, out, settings, inheritedAttributes) {
+      private val maxPendingAck = settings.bufferSize
+
+      private[jms] def createSession(connection: Connection, createDestination: Session => javax.jms.Destination) = {
+        val session =
+          connection.createSession(false, settings.acknowledgeMode.getOrElse(AcknowledgeMode.ClientAcknowledge).mode)
+        new JmsAckSession(connection, session, createDestination(session), settings.bufferSize)
+      }
+
+      private[jms] def pushMessage(msg: AckEnvelope): Unit = push(out, msg)
+
+      override private[jms] def onSessionOpened(jmsSession: JmsSession): Unit =
+        jmsSession match {
+          case session: JmsAckSession =>
+            session.createConsumer(settings.selector).onComplete {
+              case Success(consumer) =>
+                consumer.setMessageListener(new MessageListener {
+
+                  var listenerStopped = false
+
+                  def onMessage(message: Message): Unit = {
+
+                    @tailrec
+                    def ackQueued(): Unit =
+                      OptionVal(session.ackQueue.poll()) match {
+                        case OptionVal.Some(action) =>
+                          try {
+                            action()
+                            session.pendingAck -= 1
+                          } catch {
+                            case _: java.lang.IllegalStateException =>
+                              listenerStopped = true
+                          }
+                          if (!listenerStopped) ackQueued()
+                        case OptionVal.None =>
+                      }
+
+                    if (!listenerStopped)
+                      try {
+                        handleMessage.invoke(AckEnvelope(message, session))
+                        session.pendingAck += 1
+                        if (session.pendingAck > maxPendingAck) {
+                          val action = session.ackQueue.take()
+                          action()
+                          session.pendingAck -= 1
```

Comment:
@akara will this work for `CLIENT_ACKNOWLEDGE`? As it acknowledges all the messages on the session shouldn't this be reset to 0?

Reply:
@akara Or thinking about it for `CLIENT_ACKNOWLEDGE` should it only be possible to set `maxPendingAck` to 0?

Reply:
This PR is already merged, please open a new ticket for issues or use some other channel (gitter or akka-user mailing list) if it is more of a discussion/question, thanks!

Reply:
Ah sorry was just looking at the history and didn't think it's probably a strange communication channel! I'll get on gitter or create an issue if I find something wrong. Thanks.